plugins/amqp_trx_plugin/amqp_trx_plugin.cpp

```diff
@@ -0,0 +1,169 @@
+#include <eosio/amqp_trx_plugin/amqp_trx_plugin.hpp>
+#include <eosio/amqp_trx_plugin/amqp_handler.hpp>
+#include <eosio/amqp_trace_plugin/amqp_trace_plugin.hpp>
+#include <eosio/chain_plugin/chain_plugin.hpp>
+
+#include <eosio/chain/exceptions.hpp>
+#include <eosio/chain/transaction.hpp>
+#include <eosio/chain/thread_utils.hpp>
+
+#include <boost/signals2/connection.hpp>
+
+namespace {
+
+static appbase::abstract_plugin& amqp_trx_plugin_ = appbase::app().register_plugin<eosio::amqp_trx_plugin>();
+
+constexpr auto def_max_trx_in_progress_size = 100*1024*1024; // 100 MB
+
+} // anonymous
+
+namespace eosio {
+
+using boost::signals2::scoped_connection;
+
+struct amqp_trx_plugin_impl : std::enable_shared_from_this<amqp_trx_plugin_impl> {
+
+   chain_plugin* chain_plug = nullptr;
+   amqp_trace_plugin* trace_plug = nullptr;
+   // use thread pool even though only one thread currently since it provides simple interface for ioc
+   std::optional<eosio::chain::named_thread_pool> thread_pool;
+   std::optional<amqp> amqp_trx;
+
+   std::string amqp_trx_address;
+   std::atomic<uint32_t> trx_in_progress_size{0};
+
+   // called from amqp thread
+   bool consume_message( const char* buf, size_t s ) {
+      try {
+         fc::datastream<const char*> ds( buf, s );
+         fc::unsigned_int which;
+         fc::raw::unpack(ds, which);
+         if( which == fc::unsigned_int(transaction_msg::tag<chain::packed_transaction_v0>::value) ) {
+            chain::packed_transaction_v0 v0;
+            fc::raw::unpack(ds, v0);
+            auto ptr = std::make_shared<chain::packed_transaction>( std::move( v0 ), true );
+            handle_message( std::move( ptr ) );
+         } else if ( which == fc::unsigned_int(transaction_msg::tag<chain::packed_transaction>::value) ) {
+            auto ptr = std::make_shared<chain::packed_transaction>();
+            fc::raw::unpack(ds, *ptr);
+            handle_message( std::move( ptr ) );
+         } else {
+            FC_THROW_EXCEPTION( fc::out_of_range_exception, "Invalid which ${w} for consume of transaction_type message", ("w", which) );
+         }
+         return true;
+      } FC_LOG_AND_DROP()
+      return false;
+   }
+
+private:
+
+   // called from amqp thread
+   void handle_message( chain::packed_transaction_ptr trx ) {
+      const auto& tid = trx->id();
+      dlog( "received packed_transaction ${id}", ("id", tid) );
+
+      auto trx_in_progress = trx_in_progress_size.load();
+      if( trx_in_progress > def_max_trx_in_progress_size ) {
+         wlog( "Dropping trx, too many trx in progress ${s} bytes", ("s", trx_in_progress) );
+         if( trace_plug ) {
+            std::string err = "Dropped trx, too many trx in progress " + std::to_string( trx_in_progress ) + " bytes";
+            trace_plug->publish_error( trx->id().str(), chain::tx_resource_exhaustion::code_enum::code_value, std::move(err) );
```

Comment:
Is there anything that forces the `amqp_trace_plugin` to be initialized before `amqp_trx_plugin`?